### ISSUE TITLE
math tagging and security

### ISSUE THREAD

**Issue by hpvd** [latex3/tagging-project#708]
I have watched your great talk about progress of pdf tagging at TUG 2024
https://www.youtube.com/watch?v=b70CsLh7qU0

You are describing the accessibility tagging for math works via attaching small files including mathml inside.

I'm just wondering about the security topic.
Files in / attached to pdf are generally known to be a potential security risk because they can contain malicious content, open other dangerous files, or even launch applications.
Some companies therefore clean all received pdf before open them...
which damage accessibility and positive findings maybe also lead to blacklisting and block the pdf file...

=> Having this in mind , is there anything one can do in this field?

Are there other ways to make math accessible?
Or can one think of things to tightly couple pdf and mathml files e.g. via signing them? The idea is that its a least not easily possible to exchange a "good" mathml file with a "bad" actor ...

Of course this is not a latex topic, but we may use the chance to choose the most future proof direction
- security measures taken are constantly rising and pdf-accessibility should not be a blocker for accessing a pdf :-)

**Comment by hpvd** [comment]
would it be possible and used by screenreaders when adding mathml directlin in the document within a `alternate text`?

**Comment by u-fischer** [comment]
> Are there other ways to make math accessible?

Well you can add mathml tags. But this is much more difficult to automate (probably even impossible with pdftex) and also much more difficult to adjust if needed. Also even if we manage to add such mathml tags: I have doubts that pdf consumer will use them in the near future. 

> would it be possible and used by screenreaders when adding mathml directly in the document within a alternate text property?

well you naturally can put a mathml into the alternate text, but if and how a screenreader would make use of them is not something we can decide.  Currently they imho do not parse the alternate text to detect mathml and no standard requires from them such a parsing.

> According to PDF/UA accessibility guidelines, “All mathematical expressions shall be enclosed within a Formula tag and shall have an ALT attribute.”

That is PDF/UA-1. PDF/UA-2 says

> Mathematical expressions shall be represented using either structure types defined in presentation MathML, [...] or as presentation MathML via the Associated Files mechanism [...] or both.

The idea to mark up formulas with alternate text stems from the time when math were mostly given as images. With mathml and unicode it is no longer the right way. See also my issue here https://github.com/pdf-association/pdf-issues/issues/470

**Comment by hpvd** [comment]
thanks for getting back that fast and pointing to your issue in pdf-association repository https://github.com/pdf-association/pdf-issues/issues/470

**Comment by FrankMittelbach** [comment]
If I understand you correctly you think that the Associeted Files concept offer an attack vector  by putting malicious conent into such streams that are attached as AF. I would have expected that the the PDF implementations guard against that but if not it is not something we can control or influence from LaTeX other than by putting MathML or tex source into them. Even if we don't use AF ourselves anybody could then write bad code into an AF and put it into a PDF  and thus that would need to me taken up with the PDF Association (PDFA) perhaps, but for that one would probably have to prove first that there is a possibility to misuse the concept in the first place. Anyway, I don't think that it is something that  affects what we do --- if there is an attack possibility it should, of course, gets fixed but once that happened any of the approaches we consider would remain a viable possibility.

**Comment by hpvd** [comment]
yep having files within a pdf is a well known security problem.
Of course this not a latex topic.

My point is, that because of this,
people/companies take measures against this:
>Some companies therefore clean all received pdfs before open them...
which damage accessibility and positive findings maybe also lead to blacklisting and block the pdf file...

This where pdfs with this kind of accessibility features maybe damaged or sorted out.

So the idea/question is: cant we have accessibility for math formuals in a way without opening the possibility to integrate downsides like the chance of being branded as security risk

**Comment by hpvd** [comment]
Another field where not "if" there is an embedded file but "how" it is embedded seems to be important:

> **5 Risks of Associated Files**
> ....
> Usually, archives only accept PDF/A-3 files where they exercise some degree of control or
> awareness regarding the embedded files
> 
> PDF/A-3 minimizes risks in archival workflows by requiring that any embedded file is an
> Associated file associated with a limited set of objects in the PDF document. In particular,
> each such file has a well-defined MIME-type and a clear relationship information.
> 
> **3.2 PDF/A-3’s restrictions**
> PDF/A-3, which introduced Associated Files to PDF in 2012, imposes some additional
> requirements that are not present in PDF 2.0:
> ...

https://pdfa.org/resource/pdf-2-0-application-note-002-associated-files/

**Comment by u-fischer** [comment]
> Another field where not "if" there is an embedded file but "how" it is embedded seems to be important:

We know that. I'm reading PDF-specs for more than six years now ;-).  Our files all have a mime-type and a AFrelationship key.  BTW: for pdf 2.0 you should check PDF/A-4 not PDF/A-3.

**Comment by hpvd** [comment]
Just a very last comment on the including of files for accessibility:

The user experience may be a little irritating/downgraded for the standard user.

When opening the example pdf in acrobat reader, I see (of course) all embedded files.
On mousehover I see details like size 0,00 byte ..
On click I got the standard warning about "attachments my contain viruses etc.."

When opening the file in sumatra pdf, the files appears as bookmarks (a possible bug of sumatra..)

=> all together this is not the experience normal people are used to when opening PDFs produced by latex which mostly shine with perfectly crafted properties - and this is also why some editors may love latex :-)

Accessibility is a real important topic. It would be great if every with latex newly produced PDF have this out of the box and per default.

The question is:
would everyone keep this switched on if the user experience is not en par with accessibility switched of?
I already imagine PHD Students calling their students: "what happening to your diploma thesis PDF??"

Sorry for these perhaps overly skeptical words.
I love latex for more than 30years and work in the web field with many a11 and usability topics...
If there is a way to further optimize accessibility also for math formular, its worth to be found!

<img width="783" alt="2024-09-17_21h46_17" src="https://github.com/user-attachments/assets/83331925-034b-430f-b527-49e62504f58d">

<img width="591" alt="2024-09-17_21h30_39" src="https://github.com/user-attachments/assets/6f0ccc91-6d03-4b77-a13d-116ceac9f9fd">

**Comment by davidcarlisle** [comment]
@hpvd you may prefer the PDF 1.7 PDF/UA-1 versions https://github.com/latex3/tagging-project/discussions/82  although math accessibility is generally much worse with PDF 1.x.

It is well known that at the current time pdf reader support for PDF 2.0 is sadly lacking.  An explicit aim of the project is to build up a large corpus of PDF 2.0 files to put some pressure on PDF readers to support the standard.

At the current time it's not unreasonable to use PDF 2  for testing or for specific accessibility needs and PDF 1.7 PDF/UA-1 for actual production work.  If no one builds PDF 2 files, readers will never support it, so you have to break the cycle somewhere.

With development builds of foxit and NVDA/Mathcat you get good screen reading and can navigate directly in to the math expressions from the screen reader, which is a glimpse of how it is supposed to work.

**Comment by u-fischer** [comment]
We are currently adding the files to the embedded files pane, but this is mostly for debugging and to allow people to see the mathml. There is an option to switch that off, and that is what we will probably do later by default as it doesn't make much sense to show hundreds of mathml-files in the pane.

That sumatra shows embedded files as bookmarks is clearly a sumatra issue (or feature).

**Comment by josephwright** [comment]
I have a feeling this should be turned into a discussion!
